lib/class-wp-theme-json-gutenberg.php

```diff
@@ -1059,7 +1059,11 @@ private function get_block_classes( $style_nodes ) {
 			}
 
 			if ( self::ROOT_BLOCK_SELECTOR === $selector ) {
-				$block_rules .= 'body { margin: 0; }';
+				// If the root body node has margin set in its spacing property
+				// then avoid setting a default margin.
+				if ( ! isset( $node['spacing']['margin'] ) ) {
+					$block_rules .= 'body { margin: 0; }';
+				}
```

Comment:
This is the changeset to focus on in this PR

Reply:
I was thinking why not adding that value as default in theme.json in core, so the theme can override that vales witout an if statement

Reply:
Relevant to this fix is the original PR by @youknowriad and this comment https://github.com/WordPress/gutenberg/pull/35421#discussion_r724010968

cc @ntsekouras

Reply:
>I considered whether this should be applied in the default theme.json instead but decided against because while right now the default theme.json styles applies to body, I'm wondering if it's better if they apply to .wp-site-blocks because that's the actual container of everything (imagine defining a layout style at the root level, you want it to apply there and not on body)

@youknowriad's comment makes sense but I'm wondering if changing this now would cause some themes to break. Should we handle `margin` now in `body` and consider a change like this after 5.9 🤔 ?

Setting a default in theme.json sounds good, but we also need to handle both notations `margin: 10px` and `margin:{top: 10px}`.

Reply:
My personal opinion is that this PR is a _bug fix_ and we should strive to keep it as such.

I understand there are other things to explore, namely:

1. Making this part of default `theme.json` as suggested in https://github.com/WordPress/gutenberg/pull/36958#discussion_r758487621.
2. Applying the `margin: 0` to the `.wp-site-blocks` element as per https://github.com/WordPress/gutenberg/pull/35421#discussion_r724010968.

...however the aim of this PR is to fix the bug ready for WP 5.9 Beta 2 (given that Beta 1 deadline has passed) and it seems to me - given we're in a "bug fix only" period -  we should therefore keep the scope of fixes limited.

Personally I would prefer we pursued https://github.com/WordPress/gutenberg/pull/36960 as this option retains the current `margin: 0` on the `body` which is what Themes have come to expect, but it also allows Themes to overide that setting via `theme.json`. That seems to cover the use case stated in the original Issue #36147, namely:

> If I add a margin property to the theme.json at the top level styles the style generated is overwritten by a style added by the gutenberg plugin.

Given the overhead of raising and testing Global Styles PRs (PRs in both Gutenberg and Core) it would be great if we could decide on the preferred approach in order that we can get this issues resolved with maximum efficiency. I appreciate your understanding.

The decisions to be made are:

1. Do we agreed that we should proceed with this "limited" fix scope (i.e. not pursuing other explorations)?
2. Which PR should be pursue - #36958 or #36960?

Thanks all 🙇

Reply:
>Do we agreed that we should proceed with this "limited" fix scope (i.e. not pursuing other explorations)?

I personally agree with this.

>Which PR should be pursue 

I also prefer the selector ordering PR (#36960 )

Reply:
> comment makes sense but I'm wondering if changing this now would cause some themes to break. Should we handle margin now in body and consider a change like this after 5.9 🤔 ?

As a high level principle, I like the idea of serving out good defaults but giving full control to themers to trivially style everything with theme.json, style.css or otherwise.

In this case it sounds like the risk is themes assuming `body { margin: 0; }` to be present, and this PR then removing it, is that correct?

In that case, have we considered adding this rule to create a soft default? 

```
:where(body) {
	margin: 0;
}
```

In theory that should output the existing rule but be trivially overridden by theme.json.

Reply:
>In this case it sounds like the risk is themes assuming body { margin: 0; } to be present, and this PR then removing it, is that correct?

I think it's not only about `margin`, but all the other 'root' styles to be applied to `.wp-site-blocks` and not `body`.

Reply:
@jasmussen An alternative fix is https://github.com/WordPress/gutenberg/pull/36960

Reply:
See https://github.com/WordPress/gutenberg/pull/36958#issuecomment-982549243